### PR TITLE
Fix Controller:WaterCoil chilled water temperature and humidity control

### LINKED ISSUE #7001: Improved chilled water coil humidity control
Issue overview
--------------
This new feature request seeks more precise humidity control for chilled water coils.

User notes that a simple system with a fixed max humidity ratio setpoint over-dehumidifies the supply air, because the adjusted temperature setpoint appears to simply be the dew point temperature corresponding to the humidity ratio setpoint. 

Work was done in #5336 to improve humidity controls for chilled water and DX coils. When the chilled water coil controller is set for TemperatureAndHumidity control, it lowers the dry bulb temperature setpoint as needed in order to reach the maximum humidity ratio setpoint if possible. According to the code comments, the final solution was to adjust the setpoint only once per timestep in order to avoid controller iteration errors. This request would revisit the ChW controls.

It has been considered in the past that the simulation sequence for Controller:WaterCoil should be changed to be more local to the coil instead of the current airloop solution technique in `SolveAirLoopControllers` where the controllers are simulated first, then the airloops, then repeat. Not sure if that would help here or not.

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version)
 - Version of EnergyPlus (if using an intermediate build, include SHA)
 - Helpdesk ticket number 12653

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### LINKED ISSUE #7653: Convergence problem with Controller:WaterCoil and TemperatureAndHumidity control - need more plant iterations?
Issue overview
--------------
Work on changing the order of the internal surface data (#7575) has tripped over some surprising diffs. For example, here's the cooling coil load and chiller load for the summer design day in 5ZoneWaterCooled_Baseboard, with develop and with ManageSurfaceHeatBal2. Similar behavior has been observed in 5ZoneCoolingPanelBaseboardAuto (both files have baseboard with a radiant fraction, but I'm not sure that's relevant) or maybe it is. The cooling coil loads a differ some, but the chiller load shows big diversions at night when the load is smaller.

image

On a wild guess (or goose chase), I tried increasing the minimum plant iterations to 10 (from the default of 2).
````
  ConvergenceLimits,
    1,                       !- Minimum System Timestep {minutes}
    20,                      !- Maximum HVAC Iterations
    10,                       !- Minimum Plant Iterations
    20;                       !- Maximum Plant Iterations
````

This alters the chiller load significantly in both branches.
image

Looking at just develop with the change in ConvergenceLimits:

image

### Details
Some additional details for this issue (if relevant):
 - Platform Win64
 - Version of EnergyPlus 9.3.0-d0361cca92
 - Reference #7575 

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file: testfiles\5ZoneWaterCooled_Baseboard.idf
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by mjwitte**
Pull request overview
---------------------
 - Fixes #7653
 - Refine the humidity control reset for Controller:WaterCoil with control variable TemperatureAndHumidityRatio.
 - Reinstate (and fix) the Rootfinder `CheckIncrementRoundOff` function which checks to see if the next X increment is larger than the X tolerance and ends iterations if not.
 - Expand Rootfinder `CheckIncrementRoundOff` function to check for change in Y value < tolerance.
 - Remove call to `UpdateController` when `CheckBracketRoundOff` returns true.

Diffs - CI Tests
-------------
*Italics indicate potential problem files*
| File | T&HumRat? | eso/mtr Diffs |Error Diffs | 
|:---|:---:|:---:|:---|
| 5ZoneCoolingPanelBaseboard  | Y | Big | |
| 5ZoneCoolingPanelBaseboardAuto | Y | Big | |
| 5ZoneCoolingPanelBaseboardTotalLoad | Y | Big | |
| 5ZoneCoolingPanelBaseboardVarOff | Y | Big | |
| 5ZoneElectricBaseboard | Y | Big | |
| 5ZoneWaterCooled_Baseboard | Y | Big | |
| 5ZoneWaterCooled_BaseboardScalableSizing | Y | Big | |
| 5ZoneWaterCooled_GasFiredSteamHumidifier | Y | Big | |
| 5ZoneWaterCooled_HighRHControl | Y | Big | |
| 5ZoneWaterCooled_MultizoneAverageRHControl | Y | Big | |
| 5ZoneWaterCooled_MultizoneMinMaxRHControl | Y | Big | |
| DOAS_wNeutralSupplyAir_wFanCoilUnits | Y | Big | |
| EMPD5ZoneWaterCooled_HighRHControl | Y | Big | |
| HeatPumpWaterHeater | Y | Big | |
| *HeatPumpWaterHeaterStratified* | Y | Big | New warning - SolveAirLoopControllers: Maximum iterations (50) exceeded | 
| VSHeatPumpWaterHeater | Y | Big | |
| DOASDualDuctSchool | N | Small | |
| HospitalBaseline | Y | Small | Old warnings gone! - SolveAirLoopControllers: Maximum iterations (50) exceeded |
| HospitalBaselineReheatReportEMS | Y | Small | Old warnings gone! - SolveAirLoopControllers: Maximum iterations (50) exceeded |
| HospitalLowEnergy | Y | Small | Old warnings gone! - SolveAirLoopControllers: Maximum iterations (50) exceeded |
| RefBldgHospitalNew2004_Chicago | Y |  Small | Old warnings gone! - SolveAirLoopControllers: Maximum iterations (50) exceeded |

Diffs - Defect Files
-------------
For #7653 [5ZoneWaterCooled_Baseboard-Detailed.idf.txt](https://github.com/NREL/EnergyPlus/files/4713759/5ZoneWaterCooled_Baseboard-Detailed.idf.txt)
See how the chiller flow and load jumped around at the beginning of the summer design day, now they are more smooth and stable.

image

image

For #7215 which fixed #7001 (which is when 5ZoneWaterCooled_Baseboard went haywire but nobody noticed)
[7001-DOAS_CWCoil.idf.txt](https://github.com/NREL/EnergyPlus/files/4713763/7001-DOAS_CWCoil.idf.txt) This is checking to make sure these changes don't undo the fix. Basically the same results with a little tighter control now.

image

image

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [x] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] ~If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label~
 - [ ] ~If structural output changes, add to output rules file and add OutputChange label~
 - [ ] ~If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies~

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by mjwitte** [comment]
@Myoldmopar This is much better. This is the same branch we looked at yesterday, but with one small change. Now the diffs are more manageable. A quick review of the err diffs shows most files just has slight changes in the values in certain warnings, but for some reason three heat pump water heater files show new errors. Will stay on this path for now and work on cleaning up the unwanted diffs.

File: src/EnergyPlus/HVACControllers.cc [hunk 1]
```diff
@@ -2247,13 +2247,13 @@ namespace HVACControllers {
...
                         // For humidity control tolerance, always use 0.0001 which is roughly equivalent to a 0.015C change in dewpoint
-                        if (Node(thisController.SensedNode).HumRat > (Node(thisController.SensedNode).HumRatMax + 1.0e-4)) {
+                        if (Node(thisController.SensedNode).HumRat > (Node(thisController.SensedNode).HumRatMax + 1.0e-5)) {
                             // Turn on humdity control and restart controller
                             IsConvergedFlag = false;
                             thisController.HumRatCtrlOverride = true;
                             if (thisController.Action == iReverseAction) {
                                 // Cooling coil controller should always be ReverseAction, but skip this if not
-                                RootFinder::SetupRootFinder(RootFinders(ControlNum), iSlopeDecreasing, iMethodBrent, constant_zero, 1.0e-6, 1.0e-4);
+                                RootFinder::SetupRootFinder(RootFinders(ControlNum), iSlopeDecreasing, iMethodFalsePosition, constant_zero, 1.0e-6, 1.0e-5);
```
Comment (mjwitte) [inline src/EnergyPlus/HVACControllers.cc:2256]: When TemperaturAndHumidityRatio control switch to humidity, use a tighter control tolerance and a different solution method.

File: src/EnergyPlus/RootFinder.cc [hunk 2]
```diff
@@ -439,7 +439,7 @@ namespace RootFinder {
     void IterateRootFinder(RootFinderDataType &RootFinderData, // Data used by root finding algorithm
                            Real64 const X,                     // X value of current iterate
                            Real64 const Y,                     // Y value of current iterate
-                           Optional_bool IsDoneFlag            // If TRUE indicates that the iteration should be stopped
+                           bool &IsDoneFlag                    // If TRUE indicates that the iteration should be stopped
```
Comment (mjwitte) [inline src/EnergyPlus/RootFinder.cc:442]: This argument wasn't really optional. Tripped on that during unit test development.

File: src/EnergyPlus/RootFinder.cc [hunk 3]
```diff
@@ -643,21 +643,10 @@ namespace RootFinder {
...
         // Essentially we stop the iteration if:
-        // - the increment beween successive iterates is smaller than the user-specified
-        //   tolerance for the X variables.
         // - the distance between the lower and upper bounds is smaller than the user-specified
         //   tolerance for the X variables. (USING brackets from previous iteration)
-        // BUG: Relaxed check to avoid detecting round-off in case 2 successive iterates are the same!
-        // IF ( CheckIncrementRoundOff( RootFinderData, X ) ) THEN
-        //  RootFinderData%StatusFlag = iStatusOKRoundOff
-        //  RETURN
-        // END IF
```
Comment (mjwitte) [inline src/EnergyPlus/RootFinder.cc:654]: Removed comments about unused function.

File: src/EnergyPlus/RootFinder.cc [hunk 4]
```diff
@@ -643,21 +643,10 @@ namespace RootFinder {
...
         if (CheckBracketRoundOff(RootFinderData)) {
             RootFinderData.StatusFlag = iStatusOKRoundOff;
-            RootFinderData.XCandidate = X;
-
-            // Update root finder internal data with current iterate (X,Y)
-            UpdateRootFinder(RootFinderData, X, Y);
```
Comment (mjwitte) [inline src/EnergyPlus/RootFinder.cc:660]: If the upper/lower bracket has been reduced to less than the x tolerance, no need to update the X candidate again, just be done.

File: src/EnergyPlus/RootFinder.cc [hunk 5]
```diff
@@ -1375,65 +1364,6 @@ namespace RootFinder {
         return CheckRootFinderConvergence;
     }
 
-    bool CheckIncrementRoundOff(RootFinderDataType const &RootFinderData, // Data used by root finding algorithm
```
Comment (mjwitte) [inline src/EnergyPlus/RootFinder.cc:1378]: Removed unused function.

File: src/EnergyPlus/SimAirServingZones.cc [hunk 6]
```diff
@@ -3021,8 +3021,9 @@ namespace SimAirServingZones {
                                 ErrEnvironmentName = EnvironmentName;
                                 const auto CharErrOut = fmt::to_string(MaxIter);
                                 ShowWarningError("SolveAirLoopControllers: Maximum iterations (" + CharErrOut + ") exceeded for " +
-                                                 PrimaryAirSystem(AirLoopNum).Name + ", at " + EnvironmentName + ", " + CurMnDy + ' ' +
-                                                 CreateSysTimeIntervalString());
+                                                 PrimaryAirSystem(AirLoopNum).Name + ", " +
+                                                 PrimaryAirSystem(AirLoopNum).ControllerName(AirLoopControlNum) + ", at " + EnvironmentName + ", " +
```
Comment (mjwitte) [inline src/EnergyPlus/SimAirServingZones.cc:3025]: Add controller name to this error message to help pinpoint problems.

File: tst/EnergyPlus/unit/RootFinder.unit.cc [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+//#include <EnergyPlus/ScheduleManager.hh>
+//#include <EnergyPlus/SetPointManager.hh>
+//#include <EnergyPlus/SimAirServingZones.hh>
+//#include <EnergyPlus/WaterCoils.hh>
+
+#include "Fixtures/EnergyPlusFixture.hh"
+
+using namespace EnergyPlus;
+
+TEST_F(EnergyPlusFixture, RootFinder_CheckConvergence)
```
Comment (mjwitte) [inline tst/EnergyPlus/unit/RootFinder.unit.cc:75]: 2 new unit tests.

**Review by mjwitte — COMMENTED** [review]
Walkthrough:

**Comment by Myoldmopar** [comment]
This is a very nice, succinct PR, with good unit tests and excellent documentation of the changes made.  As I start reviewing, the first thing I see is the table of diffs at the top.  The only file in italics is HeatPumpWaterHeaterStratified.  Does this indicate I should start by looking closely at the results of this file first?

CI is showing diffs that look to be in agreement with your results table, nothing widespread, so that is nice.

I'll take a look at the code now.

File: src/EnergyPlus/HVACControllers.cc [hunk 8]
```diff
@@ -2247,13 +2247,13 @@ namespace HVACControllers {
                     // For temperature and humidity control, after temperature control is converged, check if humidity setpoint is met
                     if (!thisController.HumRatCtrlOverride) {
                         // For humidity control tolerance, always use 0.0001 which is roughly equivalent to a 0.015C change in dewpoint
```
Comment (Myoldmopar) [inline src/EnergyPlus/HVACControllers.cc:2249]: This comment line should be updated as it is now out of date.

File: src/EnergyPlus/HVACControllers.cc (hunk 1 above)
Comment (Myoldmopar) [reply src/EnergyPlus/HVACControllers.cc:2256]: Can you summarize the difference between iMethodBrent and iMethodFalsePosition?

File: src/EnergyPlus/RootFinder.cc (hunk 2 above)
Comment (Myoldmopar) [reply src/EnergyPlus/RootFinder.cc:442]: Yay, less optionals!

File: src/EnergyPlus/RootFinder.cc (hunk 4 above)
Comment (Myoldmopar) [reply src/EnergyPlus/RootFinder.cc:660]: That makes great sense.

File: src/EnergyPlus/RootFinder.cc (hunk 5 above)
Comment (Myoldmopar) [reply src/EnergyPlus/RootFinder.cc:1378]: Yay!

File: tst/EnergyPlus/unit/RootFinder.unit.cc [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+    EXPECT_EQ(rootFinderData.StatusFlag, DataRootFinder::iStatusNone);
+    EXPECT_FALSE(isDone);
+
+    // Set 5th point in between upper and lower, but non-convergent so it gets to the Increment check)
+    xValue = rootFinderData.XCandidate;
+    yValue = 5.0;
+    RootFinder::IterateRootFinder(rootFinderData, xValue, yValue, isDone);
+    EXPECT_EQ(rootFinderData.StatusFlag, DataRootFinder::iStatusOKRoundOff);
+    EXPECT_TRUE(isDone);
+}
```
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/RootFinder.unit.cc:176]: It's really great to see unit tests that directly test the solver tools like this.  So important, and amazing that this is the first set of tests added for RootFinder.cc.  :+1:

**Review by Myoldmopar — COMMENTED** [review]
Minor notes.  Overall this looks great.  Comment lines should be updated where appropriate.

This PR is adding a net of 104 lines of code, but the unit test addition itself is 177 lines.  So not counting the unit test, this PR is a net reduction of 73 lines of code.  Fantastic!

**Comment by Myoldmopar** [comment]
Your dramatic dark plots in the top comment really tell a great story, showing the improved control.  It is really nice.  I am building this branch locally and will do some checking.

File: src/EnergyPlus/HVACControllers.cc (hunk 1 above)
Comment (mjwitte) [reply src/EnergyPlus/HVACControllers.cc:2256]: From the function comments:
```
FalsePositionMethod
        // This function computes the next iterate using the false position method (aka regula falsi).
        // If new iterate does not lie within the lower and upper points then
        // the Bisection method is used instead.

BrentMethod
        // This function computes the next iterate using the Brent's method.
        // If new iterate does not lie within the lower and upper points then
        // the secant method is used instead.
        // Inverse quadratic interpolation using the last 3 best iterates.
        // The next root estimate is x = B + P/Q whereby B is the current best estimate
        // of the root.
```
Switching methods here was a shot in the dark that helped settle things down.  I chose false position (regula falsi) because that's a familiar method from other parts of the code. It may be worth exploring whether a method other than `iMethodBrent` would be a better overall default for Controller:WaterCoil.

**Comment by rraustad** [comment]
So the issue here is a difference in coil + pump and chiller flow/load. This branch certainly improves control. I would add plant exit vs coil water inlet temp just to round out the comparison but that would be more for a follow up to see if coil vs plant flow could be further improved with an ideal result of them being identical within a tolerance. Those flow/loads difference spikes seems like they would point to a plant convergence problem (is pump flow constant here?). I agree, this looks much better.

File: tst/EnergyPlus/unit/RootFinder.unit.cc (hunk 9 above)
Comment (mjwitte) [reply tst/EnergyPlus/unit/RootFinder.unit.cc:176]: Yes, and this could be expanded, but this file has fairly good coverage already (other than error paths) as a side-effect of other unit tests. There's also a couple of parallel universes for this with multiple similar methods in [TempSolveRoot](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/TempSolveRoot.cc) and [General::SolveRoot](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/General.cc#L124). I found these during a side-track attempt to do away with Controller:WaterCoil and make `RootFinder` obsolete. Some day . . .

**Comment by mjwitte** [comment]
That file uses `LeavingSetpointModulated` chiller flow control with a variable flow pump.

**Comment by rraustad** [comment]
It's interesting then that there is bypass flow. Especially at the end of the day where bypass flow is greater than coil flow. I guess chiller flow dominates and I wonder what a real plant would do. Chiller flow and coil flow never match and it makes me wonder if there are chiller inputs that could be changed to minimize the difference in flow.

**Comment by mjwitte** [comment]
As @EnergyArchmage has suggested, plant and chiller flow control may need some additional work. It does seem odd that the flows don't line up. Not sure what is driving that, but it's consistent here. It may be related to the chiller's gpm/ton vs the coil characteristics.

File: src/EnergyPlus/HVACControllers.cc (hunk 1 above)
Comment (Myoldmopar) [reply src/EnergyPlus/HVACControllers.cc:2256]: :+1: thanks for the clarification.

**Comment by Myoldmopar** [comment]
:+1: 

This is good to go!  Control is improved, warnings are gone, unused functions are gone, optional argument is not optional, better unit tests.  This is great!

Thanks for these changes @mjwitte.  Thanks for the thoughts, commentary and reviewing @rraustad @EnergyArchmage.

**Comment by Myoldmopar** [comment]
Pulled develop in locally and rebuilt.  Everything is fine and unit tests run fine.  This is going in.  I'll open up a follow-on issue to track the new SolveAirLoopControllers warning in HeatPumpWaterHeaterStratified, as well as any that might have already existed in other files.
